### DISCUSSION TITLE
SD CARD write and read success.  but data not match.

### DISCUSSION THREAD

**Discussion by maniemb10** [polarfire-soc/polarfire-soc-documentation#391] [Bare metal embedded software]
Hello
I'm working on SD card. 
using mss configurator i configured
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/59552685/b92667cf-c444-4134-960c-4b91824d71df)

Output drive for all SD pins set to 10mA

the code i am using polarfire soc example code.
when i am writing and reading the status is getting success but write, read buffer is not matching and read buffer always getting '0'

i added little delay (up to 1 second) between write and read-block. still i am getting the same issue 

please help me to come out from this issue. 

Thanks
Manikandan

**Comment by padmaraob** [comment]
The "VOLT PORTS" unused in the PFSOC MSS Configurator for SD, use BANK4 I/O's for VOLT PORTS. refer the [icicle-design](https://github.com/polarfire-soc/icicle-kit-reference-design) for SD configuration

**Reply by maniemb10** [reply]
thanks for the response.
i am working on our own design not icicle kit and Chip set used in our design is mpfs095
this is our schematic connection for micro SD card 
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/59552685/75b4ad86-d309-451b-9464-925e53a50f91)
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/59552685/4fffc447-3367-4632-a4bc-11336c6acb5a)

In our code we used this SD configuration 
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/59552685/d763c0fa-de49-4625-8476-759439a8ab4e)
the issue is SD CARD write and read get success in debug. but tx data, rxdata not match. rxdata is always zero.
can u please suggest us to come out from this issue? 

thanks 
Manikandan

**Reply by padmaraob** [reply]
The "SD_LT_SEL"(K7) pin from MSS Bank4 is connected with SEL(E3) of NVT4857 in your schematic.
SEL=High - 1.8V Operation
SEL=Low - 3V Operation

So use VOLT PORTS with BANK4 and enable "VOLT_SEL is inverted"

**Reply by maniemb10** [reply]
thank for the quick response.
we enabled VOLT PORTS with BANK4 and set VOLT_SEL is inverted. 
After configuration we are getting same issue.
one more thing if we configure sdr50 and 50MHz we are getting error in "mmc_init_sdcard".  we configured VOLT_SEL pin as high so it is selected to 1.8 volt operation.
please support us to come out from this issue.

thanks
Manikandan

**Reply by padmaraob** [reply]
The MMC driver is taken care of VOLT SEL, no need to configure separately.
Also check in MSS Configurator for  BANK4 I/O's voltage setting VDDI = 1.8V at "MSS I/O Attributes",  set "Register Pull" UP and try to set Out put drive strength to 8mA.

If you are using only SD in the design then can comment  below code in the example project
/*
        ASSERT(mss_does_xml_ver_support_switch() == true)

        if (switch_mssio_config(SD_MSSIO_CONFIGURATION) == false)
        {
            while(1u);
        }
        switch_external_mux(SD_MSSIO_CONFIGURATION);
*/

**Reply by maniemb10** [reply]
thanks for the response.
MSS Configurator BANK4 voltage is already set 1.8volt. 
We already configured register pull up output current to 8mA and commented "switch_external_mux()" function. still, we are in the same issue (SD CARD write and read get success in debug. but tx data, rxdata not match. rxdata is always zero).

we are tested with new SanDisk SD card without formatting and one more with fat32 formatting. both are giving same result.
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/59552685/4abe0e9a-00fc-427f-b9d7-d9fb927696fd)
This card is using for our testing.
here i am attaching debug log
![image](https://github.com/polarfire-soc/polarfire-soc-documentation/assets/59552685/023608ee-2d0b-4ca0-bd05-56fb23436bb0)
please help us to debug further.
we used UART to print debug log.
let me know if you need any other clarifications.

Thanks
Manikandan

**Reply by padmaraob** [reply]
The SD command path is clear but data bus path is not based on your debug log. The data bus is sampling the data with '0' on read operation. Can you check the data lines[0-3] state on read operation?

Can you also check what is the SCR register value  and phy_training_mmc() in the MMC driver(MSS_MMC_init() -> sdcard_oper_config() -> MSS_MMC_single_block_read(READ_SEND_SCR, scr_reg) -> phy_training_mmc())?

**Reply by maniemb10** [reply]
We are using CRO to check the data line [0-3]. The data value is always zero while we are writing and reading.
 SCR register value we got two 32bit value
1 => 0x00000000
2 => 0x00000001

we created the SD application project with MSS configurator, and we add the xml file to the soft console project.
can you please clarify us is there any use of Libro in this project build?

Thanks 
Manikandan

**Reply by padmaraob** [reply]
Update Libero design also when update configuration in the PFSoC MSS configurator

**Comment by maniemb10** [comment]
thanks padmaraob . 

VOLT PORTS pins are not connected to our SD card in our schematic.

we got new SD card and tested with that even still the issue is remained.

thanks
Manikandan

**Comment by gsriker77** [comment]
Hi Shivamani,
If you need any further assistance, you can raise case under microchip portal as below. We are glad to assist you further on it.
https://microchipsupport.force.com/s/

Thanks,
